cgmes/cgmes-conversion/src/main/java/com/powsybl/cgmes/conversion/SubstationIdMapping.java

```diff
@@ -189,33 +218,47 @@ private void addEnds(Map<String, Set<String>> substationAdjacency, PropertyBags
-            String subi = substationsIds.get(i);
+            addAdjacencyThroughTransformerEnds(substationAdjacency, sub0, substationsIds.get(i));
+        }
+    }
+
+    private void addAdjacencyThroughBranch(Map<String, Set<String>> fictitiousVoltageLevelAdjacency,
+                                           PropertyBag equipment, boolean fictitiousVoltageLevelByNode) {
+
+        CgmesTerminal t1 = context.cgmes().terminal(equipment.getId(CgmesNames.TERMINAL + 1));
+        CgmesTerminal t2 = context.cgmes().terminal(equipment.getId(CgmesNames.TERMINAL + 2));
+
+        Optional<String> nodeId1 = context.cgmes().node(t1, context.nodeBreaker());
+        Optional<String> nodeId2 = context.cgmes().node(t2, context.nodeBreaker());
+        if (nodeId1.isPresent() && !context.boundary().containsNode(nodeId1.get()) && nodeId2.isPresent() && !context.boundary().containsNode(nodeId2.get())) {
+            Optional<CgmesContainer> cgmesContainer1 = context.cgmes().nodeContainer(nodeId1.get());
+            Optional<CgmesContainer> cgmesContainer2 = context.cgmes().nodeContainer(nodeId2.get());
 
-            if (sub0.contentEquals(subi)) {
-                continue;
+            if (cgmesContainer1.isPresent() && cgmesContainer2.isPresent() && isValidReference(cgmesContainer1.get(), cgmesContainer2.get())) {
+                String voltageLevelId1 = findVoltageLevelAndRecordItIfItIsFictitious(cgmesContainer1.get(), nodeId1.get(), fictitiousVoltageLevelByNode);
+                String voltageLevelId2 = findVoltageLevelAndRecordItIfItIsFictitious(cgmesContainer2.get(), nodeId2.get(), fictitiousVoltageLevelByNode);
+                addAdjacencyThroughTransformerEnds(fictitiousVoltageLevelAdjacency, voltageLevelId1, voltageLevelId2);
             }
-            addAdjacency(substationAdjacency, sub0, subi);
         }
     }
 
-    // Record in the adjacency Map that "id1 is adjacent to id2" and "id2 is adjacent to id1"
-    private static void addAdjacency(Map<String, Set<String>> adjacency, String id1, String id2) {
-        Set<String> ad1 = adjacency.get(id1);
-        if (ad1 == null) {
-            throw new PowsyblException("Unexpected reference to Substation or voltageLevel " + id1
-                + ". It has not been defined in CGMES substations / voltageLevels.");
-        }
-        Set<String> ad2 = adjacency.get(id2);
-        if (ad2 == null) {
-            throw new PowsyblException("Unexpected reference to Substation or voltageLevel " + id2
-                + ". It has not been defined in CGMES substations / voltageLevels.");
+    // one or both of them must be LineContainers
+    private static boolean isValidReference(CgmesContainer cgmesContainer1, CgmesContainer cgmesContainer2) {
+        return !(cgmesContainer1.isVoltageLevel() && cgmesContainer2.isVoltageLevel());
+    }
+
+    private static void addAdjacencyThroughTransformerEnds(Map<String, Set<String>> adjacency, String id1, String id2) {
```

Comment:
There has been a little confusion, some renaming has been done for incorrect methods:
addAdjacencyThroughTransformerEnds should actually be: addAdjacency

Reply:
Fixed